Avoid trying to import platforms that do not exist

bdraco opened the pull request:
~~needs https://github.com/home-assistant/core/pull/112071~~

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Avoid trying to import platforms that cannot exist.  This one makes a big difference in decreasing startup time.  Numbers are down ~18% on my test Green

We can avoid 1000s of calls to `importlib` at startup time for integration platforms that do not exist. Since we are doing this in the event loop, its much better to `stat` the two place we expect the file/dir than hand it to `importlib` to `stat` every possible combination in `sys.path` only to raise `ModuleNotFound`

> [!note]
> I've been ignoring this **red hot** area of the code in profiles for so long because I've been waiting for https://github.com/python/cpython/pull/92041/ https://github.com/python/cpython/pull/92042 to get solved in cpython since it makes the profile harder to break down. While we still need to solve that issue, avoiding the import machinery altogether goes a long way to improving the status quo.




```
2024-03-01 20:21:23.226 DEBUG (MainThread) [homeassistant.loader] Importing homeassistant.components.sensorpro.diagnostics
2024-03-01 20:21:23.226 DEBUG (MainThread) [homeassistant.loader] Importing homeassistant.components.sensorpro.diagnostics took 0.00s
2024-03-01 20:21:23.226 DEBUG (MainThread) [homeassistant.loader] Importing homeassistant.components.sensorpro.recorder
2024-03-01 20:21:23.226 DEBUG (MainThread) [homeassistant.loader] Importing homeassistant.components.sensorpro.recorder took 0.00s
2024-03-01 20:21:23.226 DEBUG (MainThread) [homeassistant.loader] Importing homeassistant.components.sensorpro.media_source
2024-03-01 20:21:23.226 DEBUG (MainThread) [homeassistant.loader] Importing homeassistant.components.sensorpro.media_source took 0.00s
2024-03-01 20:21:23.226 DEBUG (MainThread) [homeassistant.loader] Importing homeassistant.components.sensorpro.system_health
2024-03-01 20:21:23.226 DEBUG (MainThread) [homeassistant.loader] Importing homeassistant.components.sensorpro.system_health took 0.00s
2024-03-01 20:21:23.226 DEBUG (MainThread) [homeassistant.loader] Importing homeassistant.components.sensorpro.group
2024-03-01 20:21:23.226 DEBUG (MainThread) [homeassistant.loader] Importing homeassistant.components.sensorpro.group took 0.00s
2024-03-01 20:21:23.226 DEBUG (MainThread) [homeassistant.loader] Importing homeassistant.components.sensorpro.logbook
2024-03-01 20:21:23.226 DEBUG (MainThread) [homeassistant.loader] Importing homeassistant.components.sensorpro.logbook took 0.00s
```

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

bdraco committed:
Avoid trying to import platforms that do not exist

bdraco committed:
adjust

bdraco committed:
fixes

bdraco committed:
cleanup

bdraco committed:
cleanup

bdraco committed:
cleanup

bdraco commented on homeassistant/loader.py:
```diff
@@ -991,15 +1025,15 @@ def _load_platform(self, platform_name: str) -> ModuleType:
         """
         full_name = f"{self.domain}.{platform_name}"
         cache: dict[str, ModuleType] = self.hass.data[DATA_COMPONENTS]
+
```
```suggestion
```

bdraco commented on homeassistant/loader.py:
```diff
@@ -991,15 +1025,15 @@ def _load_platform(self, platform_name: str) -> ModuleType:
-                missing_platforms_cache: dict[str, ImportError] = self.hass.data[
-                    DATA_MISSING_PLATFORMS
-                ]
+                missing_platforms_cache: dict[str, ImportError]
+                missing_platforms_cache = self.hass.data[DATA_MISSING_PLATFORMS]
```
```suggestion
                missing_platforms_cache: dict[str, ImportError] = self.hass.data[
                    DATA_MISSING_PLATFORMS
                ]
```

bdraco committed:
Apply suggestions from code review

bdraco committed:
docs

bdraco commented on homeassistant/loader.py:
```diff
@@ -979,6 +980,39 @@ def get_platform(self, platform_name: str) -> ModuleType:
             return platform
         return self._load_platform(platform_name)
 
+    def get_integration_platform(self, platform_name: str) -> ModuleType:
```
Maybe the name isn't so great here

balloob replied:
Why make this a new method and not put it in the normal load platform method ?

bdraco replied:
The likely outcome here is a failure to load where-as the normal load platform has a likely outcome of success

bdraco replied:
I've redesigned this

bdraco committed:
fixes

bdraco committed:
fixes

bdraco committed:
comment

bdraco committed:
coverage

bdraco committed:
coverage

bdraco committed:
coverage

bdraco committed:
Switch config to use async_get_component

This was another path where integrations that were marked to load in the executor
would be loaded in the loop

bdraco committed:
Switch config to use async_get_component/async_get_platform

This was another path where integrations that were marked to load in the executor
would be loaded in the loop

bdraco commented:
https://github.com/home-assistant/core/pull/112070 showed we need an async version that is run in the executor

bdraco commented:
I need to have two versions, other option make integration platforms use tasks instead of callbacks. Maybe not a problem that we can do that eagerly since most won't suspend anyways

bdraco commented:
Trade offs 🙈

bdraco commented:
Maybe we make this platform exists instead. And it can return True False or None if we don't know

bdraco commented:
That might be nicer as it means he common path isn't to raise an which we know is slower

bdraco commented:
Even with eager 4000 tasks is still bad. I guess we need to go to pre-checkout.

bdraco commented:
We do the pre-check then we can switch to await and config, and we use the async loader for platforms

bdraco commented:
That also gets rid of the concept of integration platform in the loader. I like that better

bdraco committed:
Merge branch 'dev' into avoid_import_for_platforms

bdraco committed:
merge

bdraco committed:
Merge branch 'config_loads_loop' into avoid_import_for_platforms

bdraco committed:
refactor

bdraco committed:
refactor

bdraco committed:
Merge remote-tracking branch 'upstream/avoid_import_for_platforms' into avoid_import_for_platforms

bdraco committed:
coverage

bdraco committed:
preen

bdraco committed:
preen

bdraco committed:
Merge branch 'dev' into avoid_import_for_platforms

MartinHjelmare commented on homeassistant/config.py:
```diff
@@ -1432,22 +1432,24 @@ async def async_process_component_config(  # noqa: C901
-            )
-            config_exceptions.append(exc_info)
-            return IntegrationConfigInfo(None, config_exceptions)
+    if integration.platform_exists("config") is not False:
```
This call is blocking.

bdraco replied:
Yes it is, but its far cheaper than what importlib will do if its called so its worth the trade off. There isn't a currently isn't a good solution that doesn't flood the executor with call to do the two `stat`s.

bdraco replied:
Ideally we come up with a way to pre do the `stat()` calls in a single executor job, but we don't know the location of the component on disk until we load it and can check `__file__` so its not simple to solve

MartinHjelmare replied:
Please always add a code comment when there's an exception to our normal procedures, and explain why we do it that way. In this case the normal procedure would be to schedule the blocking call on the executor.

bdraco replied:
I think I can make this particular one go away in two more turns after https://github.com/home-assistant/core/pull/112096

If I can't get rid of it soon, I'll add comments

The one in integrtion_platform is harder to solve so I'll add that one now

bdraco replied:
https://github.com/home-assistant/core/pull/112103

bdraco replied:
> I think I can make this particular one go away in two more turns after #112096

https://github.com/home-assistant/core/pull/112104 will add tests + more comments after sleep as its past midnight here

bdraco replied:
https://github.com/home-assistant/core/pull/112168 will eliminate the I/O in the event loop when merged along with https://github.com/home-assistant/core/pull/112127

bdraco commented on homeassistant/helpers/integration_platform.py:
```diff
@@ -48,6 +48,10 @@ def _get_platform(
         )
         return None
 
+    if integration.platform_exists(platform_name) is False:
```
The one here is called significantly more (order of magnitude more at least). If we get to `integration.get_platform` the amont of blocking I/O it will do is two or three orders of magnitude more than the `stat()` calls
